### DISCUSSION TITLE
Axis Dewarp live View

### DISCUSSION THREAD

**Discussion by rajesh-venigalla** [AxisCommunications/discussions#186] [Axis Windows SDKs] [answered]
Hi,

I downloaded Axis Dewarping SDK and I could use it to Dewarp recorded video. 

I am trying to find a way to view live from 180 degree camera using the same Axis Dewarp SDK but could not find a sample.

I appreciate if any one can help me with a link to a sample SDK or code to view Dewarp live from 180 Axis Camera. (Axis P3813-PVE)

Rajesh

**Comment by fre3de** [comment]
Hi,

The sample VirtualPtzP3807 included in the Dewarp SDK shows how to view live view from a camera with Dewarp sdk. 
Do you use P3818 camera? We have made improvements to the Dewarp SDK for 3818 and 3819. If you come across problems with the virtual ptz movments we can send you an updated version of the videotransform.dll.

Best regards
Fredrik

**Reply by rajesh-venigalla** [reply]
Fredrik,

Thank you for responding. I appreciate it.
Yes, I use P3818 camera.
I see exceptions related to video transform. I would like to try the updated version of videotransform.dll

Thank you for your help in advance.

Regards,
Rajesh

**Reply by fre3de** [reply]
What kind of exceptions do you get, can you show the stack trace?
P3818 and P3819 at high resolution may not work correctly with the AxisRtpSrcFilter.ax file included in AxisMediaParser SDK. 
I have attached the latest versions of the two files. Copy the 64-bit versions to C:\Program Files\Axis Communications\Components and the 32-bit versions  to C:\Program Files (x86)\Axis Communications\Components.
You find the files here: https://we.tl/t-aYpfNvq4si

Best regards
Fredrik

**Reply by rajesh-venigalla** [reply]
Thank you for providing new files.

Exception Message : Transform Table API is not supported by device
Exception Stack Trace :    at CustomTransform.CustomTransformVideoForm.GetTransformData() in C:\Program Files (x86)\Axis Communications\AXIS Dewarping SDK\Samples\VirtualPtzP3807\CustomTransformVideoForm.cs:line 275
   at CustomTransform.CustomTransformVideoForm.InitDewarpingTransform(Int32 outputWidth, Int32 outputHeight) in C:\Program Files (x86)\Axis Communications\AXIS Dewarping SDK\Samples\VirtualPtzP3807\CustomTransformVideoForm.cs:line 206

I see the exception even after replacing both files in 64-bit versions to C:\Program Files\Axis Communications\Components and the 32-bit versions to C:\Program Files (x86)\Axis Communications\Components.

Resolution used in "1920*960".

I rebooted camera and computer before testing with updated files and resolution.

I am asking this question based on the exception message. Does this camera model support Dewarp. If yes, is there any setting that I need to make sure is set correctly, to make it work?

Regards,
Rajesh

**Reply by fre3de** [reply, marked as answer]
P3818 always stitches the image in the camera. No need to do client side stitching as in the example. The image from P3818 doesn't need to be dewarped but with the dewarp sdk you can do horizon calibration and get virtual ptz control. 
 
 Remove the following code from the example: 

                        byte[] transformData = null;
			try
			{
				transformData = GetTransformData();
			}
			catch (Exception ex)
			{
				System.Diagnostics.Trace.WriteLine("Failed to retrieve transform data.\n" + ex.Message);
			}

			if (transformData != null)
			{
				// Set custom transform data
				try
				{
					dewarpingTransform.SetCustomTransformData(transformData, 0);
					isClientDewarpMode = true;
				}
				catch (ArgumentException)
				{
					MessageBox.Show("Invalid transform data");
				}
			}
			else
			{
				isClientDewarpMode = false;
			}
		}

**Reply by rajesh-venigalla** [reply]
Thank you for quick response. 

I can see the live view with or without removing the code. But I cannot perform virtual PTZ. 

I would like to be able to view live from multiple sections of camera, just like I was able to view different sections in playback in the sample video "C:\Program Files (x86)\Axis Communications\AXIS Dewarping SDK\Samples\DewarpDirectShowCSharp\bin\x86\Debug\sample_overview_ceiling.H264.bin"

I attached sample playback screen shot. I would like to do the same thing in live. 

If I have to use a different camera to achieve this I can use that camera.

![Dewarp](https://user-images.githubusercontent.com/126908561/224084337-88f9677b-e768-4fb9-82d8-0b652f3dcc60.png)

**Reply by fre3de** [reply]
What happens if you zoom into the image with the mousewheel? Can you not click in the image after zooming in? Does horizon calibration work for you? 

The different views, like the quad view you show above, are for 360 cameras like M3077 and M4308 just to mention two of them.  P3818 doesn't need to be dewarped that way.

Even if you cannot use the"quad view" for P3818 you could of course do a custom solution and show the part of the image that you are interested in. One way could be to use one instance of the dewarp sdk for every view. 
In the VirtualPtz sample you see the values on zoom, pan, tilt in the top left corner.

**Reply by rajesh-venigalla** [reply]
I cannot perform zoom in or any other ptz operation on the live view using mousewheel. I can see the values of zoom, pan, tilt values on top left and they do change, however there is no change in the live view.

Nothing in the live view seems to change even when I try to calibrate using mousewheel and +/- keys, except a red horizontal line at the center.

Is there anything else that I can try to make zoom in and ptz work?

**Reply by fre3de** [reply]
Sorry for late reply.

Does the video play or do you get a frozen image? Can you try to change the video renderer. Try both AMV_RENDER_OPTIONS.AMV_VIDEO_RENDERER_EVR and AMV_RENDER_OPTIONS.AMV_VIDEO_RENDERER_VMR9
